qiskit_machine_learning/kernels/base_kernel.py

```diff
@@ -0,0 +1,133 @@
+# that they have been altered from the originals.
+
+"""Base kernel"""
+
+from __future__ import annotations
+
+from abc import abstractmethod, ABC
+
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.circuit.library import ZZFeatureMap
+
+
+class BaseKernel(ABC):
+    r"""
+    Abstract class defining the Kernel interface.
+
+    The general task of machine learning is to find and study patterns in data. For many
+    algorithms, the datapoints are better understood in a higher dimensional feature space,
+    through the use of a kernel function:
+
+    .. math::
+
+        K(x, y) = \langle f(x), f(y)\rangle.
+
+    Here K is the kernel function, x, y are n dimensional inputs. f is a map from n-dimension
+    to m-dimension space. :math:`\langle x, y \rangle` denotes the dot product.
+    Usually m is much larger than n.
+
+    The quantum kernel algorithm calculates a kernel matrix, given datapoints x and y and feature
+    map f, all of n dimension. This kernel matrix can then be used in classical machine learning
+    algorithms such as support vector classification, spectral clustering or ridge regression.
+    """
+
+    def __init__(self, feature_map: QuantumCircuit = None, enforce_psd: bool = True) -> None:
+        """
+        Args:
+            feature_map: Parameterized circuit to be used as the feature map. If None is given,
+                the `ZZFeatureMap` is used with two qubits.
+            enforce_psd: Project to closest positive semidefinite matrix if x = y.
+                Default True.
+        """
+        if feature_map is None:
+            feature_map = ZZFeatureMap(2)
+
+        self._feature_map = feature_map
+        self._enforce_psd = enforce_psd
+
+    @abstractmethod
+    def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+        r"""
+        Construct kernel matrix for given data.
+
+        If y_vec is None, self inner product is calculated.
+
+        Args:
+            x_vec: 1D or 2D array of datapoints, NxD, where N is the number of datapoints,
+                D is the feature dimension
+            y_vec: 1D or 2D array of datapoints, MxD, where M is the number of datapoints,
+                D is the feature dimension
+
+        Returns:
+            2D matrix, NxM
+        """
+        raise NotImplementedError()
+
+    def _validate_input(
+        self, x_vec: np.ndarray, y_vec: np.ndarray | None
+    ) -> tuple[np.ndarray, np.ndarray | None]:
+        x_vec = np.asarray(x_vec)
+
+        if x_vec.ndim > 2:
+            raise ValueError("x_vec must be a 1D or 2D array")
+
+        if x_vec.ndim == 1:
+            x_vec = np.reshape(x_vec, (-1, len(x_vec)))
+
+        if x_vec.shape[1] != self._feature_map.num_parameters:
+            # before raising an error we try to adjust the feature map
+            # to the required number of qubit.
+            try:
+                self._feature_map.num_qubits = x_vec.shape[1]
```

Comment:
This logic was adapted from another algorithm, perhaps VQE, where the same approach without warnings is implemented. I'd keep as is for consistency reasons.

Reply:
Seems reasonable. I just thought it would be more transparent and probably help with debugging as I think if someone uses training data that doesn't fit the feature map it's probably more likely to be a mistake. It's nice that the algorithm still runs but might lead to unexpected results..

Reply:
The idea was these were more template circuits where you defined what you wanted upfront but maybe at that time do not have detail on the problem. Later when we do have actual problem it can then be updated. In VQE it adjusts the ansatz to match the operator.